### PR TITLE
CRUD: populates `items` and `itemsCount` when a new item is created

### LINKED ISSUE #47627: CRUD: populates store when a new item is created
The key to our strategy for updating the store optimistically is understanding the data structure in advance. This allows us to populate the store precisely when a new item is created, anticipating its shape.

Let's examine the Attribute Terms store (`/wc/admin/products/attributes/terms`):

Screenshot 2024-05-20 at 15 56 33

The image above shows the status of the store. It has four items:

* `data`: contains the raw data
* `items`: links the `GET_ITEMS` indexes with the raw data used by the selectors.
* `itemsCounts`: it computes the number of items based on the data. It's used by counter-selectors

Now, let's suppose a new item is created, the response has the following shape:

```json
{
  "id": 600,
  "name": "pink",
  "slug": "pink",
  "description": "",
  "menu_order": 0,
  "count": 0,
}
```

Currently, when a new item is created, it's added to the store's `data` key. However, the `items` and `itemsCounts` keys are not updated. To address this, we need to update the implementation of these properties without triggering a new sync request (invalidate resolution).

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR populates the store's `items` and `itemsCount` properties when a new item is created. Depending on the case, it can improve the UX.
Important: the `optimistic` mode isn't enabled by default. It doesn't change the current behavior of the CRUD store.

Depends on https://github.com/woocommerce/woocommerce/issues/47623
Closes https://github.com/woocommerce/woocommerce/issues/47627

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

The PR is full of tests, trying to cover the most critical cases. Please feel free to take a look [here](https://github.com/woocommerce/woocommerce/blob/9a021263adf601a4b210b190aeb6441c70f1e538/packages/js/data/src/crud/test/utils.ts) and [here](https://github.com/woocommerce/woocommerce/blob/9a021263adf601a4b210b190aeb6441c70f1e538/packages/js/data/src/crud/test/reducer.ts).

You can run tests locally:

```cli
pnpm --filter=./packages/js/data run test:js packages/js/data/src/crud/test/
```

Screenshot 2024-05-20 at 16 52 55

You can use the client dev tools and the `data` package to test the.

1. Ensure to have installed the Redux dev tool ([Chrome](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?pli=1) / [Firefox](https://addons.mozilla.org/en-US/firefox/addon/reduxdevtools/))
2. Enable the Product editor
3. Create a new product
4. Open the dev console and the Redux dev tool
5. Filter the store by `/attributes`. Select the last one `wc/admin/products/attributes`:

Screenshot 2024-05-20 at 17 01 53

6. In the dev console, dispatch an action to create a new global attribute. The snippet below creates a new `color` attribute. 

```es6
wp.data.dispatch( 'wc/admin/products/attributes' ).createProductAttribute( {
    name: 'color',
    generate_slug: true,
},
{
        optimisticQueryUpdate: { order_by: 'name' },
} )
```

7. Confirm that the `items` and `itemsCount` populate after the attribute is created:

Screenshot 2024-05-20 at 17 04 53

8. You'd like to go to the Variations section to check the data propagation in the

image

9. Pick the attribute ID recently created for use in the following steps. In my testing, it's `886`
10. Select the `wc/admin/products/attributes/terms` store:

image

11. Dispatch a new action to create now an attribute term. Pass the attribute ID of the step 6

```es6
wp.data.dispatch( 'wc/admin/products/attributes/terms' ).createProductAttributeTerm( {
    name: 'pink',
    attribute_id: 886,
}, {
    optimisticQueryUpdate: {
        search: '',
        attribute_id: 886,
    },
    optimisticUrlParameters: [ 886 ],
} )
```

12. Confirm the `items` and `itemsCount` get populated:

image

13. You can confirm the new attribute in the options selector:

image

### 🍿 

https://github.com/woocommerce/woocommerce/assets/77539/3a0e91a6-8c60-4a62-82af-42352e394567

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/data/src/crud/reducer.ts [hunk 1]
```diff
@@ -85,23 +89,41 @@ export const createReducer = (
...
 				case TYPES.CREATE_ITEM_SUCCESS: {
-					const createItemSuccessRequestId = getRequestIdentifier(
-						CRUD_ACTIONS.CREATE_ITEM,
-						payload.key,
-						payload.query
+					const { options } = payload;
+
+					const { objItems, ids } = organizeItemsById(
+						[ payload.item ],
+						options.optimisticUrlParameters,
```
Comment (louwie17) [inline packages/js/data/src/crud/reducer.ts:96]: `options` can be undefined, this actually break the current UI as it doesn't pass options in upon creation.

Comment (louwie17) [reply packages/js/data/src/crud/reducer.ts:96]: Screenshot 2024-05-27 at 1 09 57 PM

**Review by louwie17 — CHANGES_REQUESTED** [review]
Your test instructions tested well, but the actual changes broke the application. Also I noticed that creating a shipping class doesn't work either, ~~I haven't checked with `trunk` yet, but I assume it may be related.~~ I checked with `trunk` it seems to work correctly in trunk ( check the shipping tab and shipping class dropdown to test ).
Given there are multiple data types that use the crud store, we should test all of these areas ( attributes, categories, shipping classes, shipping zones, tags, product variations, tax classes ).

**Comment by retrofox** [comment]
> Also I noticed that creating a shipping class doesn't work either, ~I haven't checked with `trunk` yet, but I assume it may be related.~ I checked with `trunk` it seems to work correctly in trunk ( check the shipping tab and shipping class dropdown to test ).

It should be related to the comment about the `undefined` value. The goal is not to add breaking changes, meaning everything should work as usual as long as the options aren't defined. Looking...

File: packages/js/data/src/crud/reducer.ts (hunk 1 above)
Comment (retrofox) [reply packages/js/data/src/crud/reducer.ts:96]: Thanks for the test and feedback 💯

Comment (retrofox) [reply packages/js/data/src/crud/reducer.ts:96]: It's a matter of setting a default value for options, [tackled here](https://github.com/woocommerce/woocommerce/pull/47632/commits/9711243c6315181ebe9b5586956d57b930958277). Also, I've defined the `options` options, which shows the TS linting error:


image

image

**Review by louwie17 — APPROVED** [review]
Thanks for the updates @retrofox, it seemed to test well this time, also shipping classes, tags, categories, and brands worked correctly 🎉

**Pull request merged**